packages/astro/src/assets/services/service.ts

```diff
@@ -0,0 +1,174 @@
+import { AstroError, AstroErrorData } from '../../core/errors/index.js';
+import { isRemotePath } from '../../core/path.js';
+import { isESMImportedImage } from '../internal.js';
+import { ImageTransform, OutputFormat } from '../types.js';
+
+export type ImageService = LocalImageService | ExternalImageService;
+
+export function isLocalService(service: ImageService | undefined): service is LocalImageService {
+	if (!service) {
+		return false;
+	}
+
+	return 'transform' in service;
+}
+
+export function parseQuality(quality: string): string | number {
+	let result = parseInt(quality);
+	if (Number.isNaN(result)) {
+		return quality;
+	}
+
+	return result;
+}
+
+interface SharedServiceProps {
+	/**
+	 * Return the URL to the endpoint or URL your image are generated from.
+	 *
+	 * For a local service, your service should expose an endpoint handling the image requests, or use Astro's at `/_image`.
+	 *
+	 * For external services, this should point to the URL where your images are coming from, for instance, `/_vercel/image`
+	 *
+	 */
+	getURL: (options: ImageTransform) => string;
+	/**
+	 * Return any additional HTML attributes separate from `src` your service require to show the image properly.
+	 *
+	 * For example, you might want to return the result `width` and `height` to avoid CLS, or a particular `class` or `style`.
+	 * In most cases, you'll want to return directly what your user supplied you, minus the attributes that were used to generate the image.
```

Comment:
Nit:

```suggestion
	 * Return any additional HTML attributes separate from `src` that your service requires to show the image properly.
	 *
	 * For example, you might want to return the `width` and `height` to avoid CLS, or a particular `class` or `style`.
	 * In most cases, you'll want to return directly what your user supplied you, minus the attributes that were used to generate the image.
```